clients/src/main/java/org/apache/kafka/clients/consumer/internals/StreamsMembershipManager.java

```diff
@@ -0,0 +1,1249 @@
+        subscriptionState.unsubscribe();
+        transitionToSendingLeaveGroup(false);
+    }
+
+    /**
+     * This will process the assignment received if it is different from the member's current
+     * assignment. If a new assignment is received, this will make sure reconciliation is attempted
+     * on the next call of `poll`. If another reconciliation is currently in process, the first `poll`
+     * after that reconciliation will trigger the new reconciliation.
+     *
+     * @param activeTasks Target active tasks assignment received from the broker.
+     * @param standbyTasks Target standby tasks assignment received from the broker.
+     * @param warmupTasks Target warm-up tasks assignment received from the broker.
+     */
+    private void processAssignmentReceived(Map<String, SortedSet<Integer>> activeTasks,
+                                           Map<String, SortedSet<Integer>> standbyTasks,
+                                           Map<String, SortedSet<Integer>> warmupTasks) {
+        replaceTargetAssignmentWithNewAssignment(activeTasks, standbyTasks, warmupTasks);
+        if (!targetAssignmentReconciled()) {
+            transitionTo(MemberState.RECONCILING);
+        } else {
+            log.debug("Target assignment {} received from the broker is equals to the member " +
+                    "current assignment {}. Nothing to reconcile.",
+                targetAssignment, currentAssignment);
+            // Make sure we transition the member back to STABLE if it was RECONCILING (ex.
+            // member was RECONCILING unresolved assignments that were just removed by the
+            // broker), or JOINING (member joining received empty assignment).
+            if (state == MemberState.RECONCILING || state == MemberState.JOINING) {
+                transitionTo(MemberState.STABLE);
+            }
+        }
+    }
+
+    private boolean targetAssignmentReconciled() {
+        return currentAssignment.equals(targetAssignment);
+    }
+
+    private void replaceTargetAssignmentWithNewAssignment(Map<String, SortedSet<Integer>> activeTasks,
+                                                          Map<String, SortedSet<Integer>> standbyTasks,
+                                                          Map<String, SortedSet<Integer>> warmupTasks) {
+        targetAssignment.updateWith(activeTasks, standbyTasks, warmupTasks)
+            .ifPresent(updatedAssignment -> {
+                log.debug("Target assignment updated from {} to {}. Member will reconcile it on the next poll.",
+                    targetAssignment, updatedAssignment);
+                targetAssignment = updatedAssignment;
+            });
+    }
+
+    /**
+     * Called by the network thread to reconcile the current and target assignment.
+     */
+    @Override
+    public NetworkClientDelegate.PollResult poll(long currentTimeMs) {
+        if (state == MemberState.RECONCILING) {
+            maybeReconcile();
```

Comment:
I moved the reconciliation to `poll` in the consumer membership manager back then, because we may receive new metadata for resolving topic UUIDs. Does it still make sense to reconcile in poll?

Reply:
I think, that is a broader discussion that I would like to postpone after merging this PR.